Spanner mutation implementations

ChengyuanZhao opened the pull request:
Added implementations for mutation factory , mapper, and unit tests

ChengyuanZhao commented:
@meltsufin @balopat PTAL

ChengyuanZhao commented on spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/core/SpannerObjectMapperImplTests.java:
```diff
@@ -0,0 +1,226 @@
+	}
+
+	@Test
+	public void writeTest() throws ClassNotFoundException {
```
This test is ensuring that the correctly typed "to" methods are called. This is vital, because that's where our responsibility leaves and it's up to the Spanner client lib. The key point of the write-mapping is to make sure the correctly typed functions are mapped.  I will rename the fields.

ChengyuanZhao commented on spring-cloud-gcp-data-spanner/src/test/java/org/springframework/cloud/gcp/data/spanner/core/SpannerMutationFactoryImplTests.java:
```diff
@@ -0,0 +1,155 @@
+		assertEquals("custom_test_table", mutation.getTable());
+		assertEquals(Op.INSERT_OR_UPDATE, mutation.getOperation());
+	}
+
```
We don't have a public Replace method as part of the interface, so I removed the REPLACE option from the private writeBuilder function.

balopat commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java:
```diff
@@ -0,0 +1,183 @@
+
+	@Override
+	public <R> R read(Class<R> type, Struct source) {
+		R object = BeanUtils.instantiateClass(type);
```
why are we using a BeanUtils instantiation? A simple no-arg constructor can be assumed for entities.

ChengyuanZhao replied:
will change to use standard Java object creation

ChengyuanZhao replied:
Check what happens if the columns in struct don't match what's in entity and vice versa for write.

balopat commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java:
```diff
@@ -0,0 +1,183 @@
+						valueSet = true;
+						break;
+					}
+					catch (IllegalAccessException e) {
```
nit: collapse catch blocks

ChengyuanZhao replied:
perhaps need to reverse the isAssignableFrom here

balopat commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java:
```diff
@@ -0,0 +1,183 @@
+
+			if (!valueSet) {
+				throw new UnsupportedOperationException(
+						"No compatible types in object for reading column: " + name);
```
nit: "object" is confusing. I'd say: "None of the fields of class " + type + " can be used for reading column " + name

balopat commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java:
```diff
@@ -0,0 +1,183 @@
+								"Could not set value for read.", e);
+					}
+					catch (InvocationTargetException e) {
+						throw new UnsupportedOperationException(
```
nit: include the type and the current method of the source in the error message

ChengyuanZhao commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerMutationFactoryImpl.java:
```diff
@@ -0,0 +1,143 @@
+
+	@Override
+	public <T> Mutation delete(Class<T> entityClass, Iterable<? extends T> entities) {
+		return deleteKeySetFromTable(entityClass, (persistentEntity) -> {
```
make a private method out of the lambda function here

ChengyuanZhao commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerMutationFactoryImpl.java:
```diff
@@ -0,0 +1,143 @@
+	}
+
+	private Mutation deleteKeySetFromTable(Class entityClass,
+			Function<SpannerPersistentEntity, KeySet> keyFunction) {
```
get rid of this function and the lambda function actually

ChengyuanZhao commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/SpannerTemplate.java:
```diff
@@ -171,4 +160,13 @@ public long count(Class entityClass) {
+
+	private <T, U> void applyMutationOfTwoArg(BiFunction<T, U, Mutation> function, T arg1,
+			U arg2) {
+		this.databaseClient.write(Arrays.asList(function.apply(arg1, arg2)));
```
rename to applyMutationWithClass

ChengyuanZhao commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapper.java:
```diff
@@ -42,5 +42,4 @@
 	 * @return A list of objects.
 	 */
 	<T> List<T> mapToUnmodifiableList(ResultSet resultSet, Class<T> entityClass);
```
not necessarily unmodifiable ? check other examples

ChengyuanZhao replied:
@balopat Hey good morning, I made changes we discussed in latest commit. PTAL !

ChengyuanZhao commented:
@balopat I added the annotation for those inner types in Iterables. I also added caching of the get-methods for reads, but not the writes. The reason is that we have a findAll method in our template , but the insert/update/upsert methods all just take one object at a time.

ChengyuanZhao commented:
@balopat caching for read and write methods and handling the Iterable added! PTAL

balopat commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java:
```diff
@@ -0,0 +1,336 @@
+			// The casting of each biconsumer below is needed for Java 8 compilation, but not Java 9.
+			new ImmutableMap.Builder<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>>()
+			.put(Boolean.class,
+					(BiConsumer<ValueBinder<WriteBuilder>, Iterable>) (binder,
```
This compilation error is so strange...
works for me when I add the elements one by one using the builder reference:  

```
	private static final Map<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>>
			writeBuilderIterableMapping =
			// The casting of each biconsumer below is needed for Java 8 compilation, but not Java 9.
			createWriteBuilderIterableMapping();

	private static ImmutableMap<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>> createWriteBuilderIterableMapping() {
		Builder<Class, BiConsumer<ValueBinder<WriteBuilder>, Iterable>> builder = new Builder<>();

		builder.put(Date.class, ValueBinder::toDateArray);
		builder.put(Boolean.class, ValueBinder::toBoolArray);
		builder.put(Long.class, ValueBinder::toInt64Array);
		builder.put(String.class, ValueBinder::toStringArray);
		builder.put(Double.class, ValueBinder::toFloat64Array);
		builder.put(Timestamp.class,ValueBinder::toTimestampArray);
		builder.put(ByteArray.class, ValueBinder::toBytesArray);

		return builder.build();
	}
```

balopat commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java:
```diff
@@ -0,0 +1,336 @@
+			.put(Double.class, (struct, colName) -> struct.getDoubleList(colName))
+			.put(Timestamp.class, (struct, colName) -> struct.getTimestampList(colName))
+			.put(Date.class, (struct, colName) -> struct.getDateList(colName))
+			.put(ByteArray.class, (struct, colName) -> struct.getBytesList(colName))
```
we can just use method reference here too

ChengyuanZhao replied:
Changes made!

ChengyuanZhao commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java:
```diff
@@ -76,7 +74,7 @@
 		builder.put(ByteArray.class, ValueBinder::toBytesArray);
 
-		writeBuilderIterableMapping = builder.build();
+		iterablePropertyType2ToMethodMap = builder.build();
```
Is that "2To" redundant ? 

Also, it looks like this commit broke the build !

balopat replied:
well, I'm mapping to a "to method"...it's awkward. Without the 2 it reads like iterablePopertyType to method map. I'll check on the build.

balopat commented on spring-cloud-gcp-data-spanner/src/main/java/org/springframework/cloud/gcp/data/spanner/core/mapping/SpannerObjectMapperImpl.java:
```diff
@@ -0,0 +1,360 @@
+				}
+				if (getMethod != null) {
+					try {
+						accessor.setProperty(property, getMethod.invoke(source, columnName));
```
We'll need to add checking for reading null columns (on this `getMethod.invoke`), as we have the [StructReader interface behave like  this](https://googlecloudplatform.github.io/google-cloud-java/0.25.0/apidocs/com/google/cloud/spanner/StructReader.html): 
"The getTypeName() methods should only be called for non-NULL values, otherwise a NullPointerException is raised; isNull(int)/isNull(String) can be used to test for NULL-ness if necessary."

ChengyuanZhao replied:
We have that right ? 

```
if (source.isNull(columnName)) {
			continue;
		}
```

balopat commented:
@ChengyuanZhao I'm wondering if we can remove bd1a7767ce32135553608b75fd993d80b7c7242a for now and raise a separate PR for it? Otherwise it's looking good, and we could merge it.
